modules/ggl-exec/src/exec.c

```diff
@@ -105,3 +117,178 @@ GglError ggl_exec_kill_process(pid_t process_id) {
+    }
+
+    // Process finished
+    case 4: {
+        siginfo_t info = { 0 };
+        int ret = -1;
+
+        while (((ret = waitid(
+                     P_PIDFD, (id_t) context->pidfd, &info, WEXITED | WNOHANG
+                 ))
+                != 0)
+               && (errno == EINTR)) { }
+
+        if (ret != 0) {
+            GGL_LOGE("Couldn't wait for process (%d).", context->pidfd);
+            return GGL_ERR_FAILURE;
+        }
+        if ((info.si_code != CLD_EXITED) || (info.si_status != 0)) {
+            GGL_LOGE(
+                "Process exited uncleanly (%d:%d).",
+                info.si_code,
+                info.si_status
+            );
+            return GGL_ERR_FAILURE;
+        }
+        GGL_LOGD("Process finished.");
+        return GGL_ERR_EXPECTED;
+    }
+
+    default:
+        return GGL_ERR_INVALID;
+    }
+}
+
+GglError ggl_exec_command_with_output(
+    const char *const args[], GglBuffer *output
+) {
+    GglByteVec vector = ggl_byte_vec_init(*output);
+    output->len = 0;
+
+    int cout_pipe[2] = { -1, -1 };
+    int cerr_pipe[2] = { -1, -1 };
+    posix_spawn_file_actions_t action = { 0 };
+
+    int ret = pipe(cout_pipe);
+    if (ret != 0) {
+        GGL_LOGE("Failed to create pipe (%d)", errno);
+        return GGL_ERR_FAILURE;
+    };
+    GGL_CLEANUP(cleanup_close, cout_pipe[0]);
+    GGL_CLEANUP(cleanup_close, cout_pipe[1]);
+
+    ret = pipe(cerr_pipe);
+    if (ret != 0) {
+        GGL_LOGE("Failed to create pipe (%d)", errno);
+        return GGL_ERR_FAILURE;
+    }
+    GGL_CLEANUP(cleanup_close, cerr_pipe[0]);
+    GGL_CLEANUP(cleanup_close, cerr_pipe[1]);
+
+    posix_spawn_file_actions_init(&action);
+    posix_spawn_file_actions_addclose(&action, cout_pipe[0]);
+    posix_spawn_file_actions_addclose(&action, cerr_pipe[0]);
+    posix_spawn_file_actions_adddup2(&action, cout_pipe[1], 1);
+    posix_spawn_file_actions_adddup2(&action, cerr_pipe[1], 2);
+
+    posix_spawn_file_actions_addclose(&action, cout_pipe[1]);
+    posix_spawn_file_actions_addclose(&action, cerr_pipe[1]);
+    int pidfd = -1;
+    ret = pidfd_spawnp(
+        &pidfd, args[0], &action, NULL, (char *const *) args, environ
+    );
+
+    if (ret != 0) {
+        GGL_LOGE("Error, unable to spawn (%d)", ret);
+        return GGL_ERR_FAILURE;
+    }
+    GGL_CLEANUP(cleanup_close, pidfd);
+
+    int epoll = -1;
+    GglError err = ggl_socket_epoll_create(&epoll);
+    if (err != GGL_ERR_OK) {
+        return err;
+    }
+
+    ret = fcntl(cout_pipe[0], O_NONBLOCK);
+    if (ret != 0) {
+        return GGL_ERR_FAILURE;
+    }
+
+    err = ggl_socket_epoll_add(epoll, cout_pipe[0], 1);
+    if (err != GGL_ERR_OK) {
+        return err;
+    }
+
+    ret = fcntl(cerr_pipe[0], O_NONBLOCK);
+    if (ret != 0) {
+        return GGL_ERR_FAILURE;
+    }
+    err = ggl_socket_epoll_add(epoll, cerr_pipe[0], 2);
+    if (err != GGL_ERR_OK) {
+        return err;
+    }
+
+    err = ggl_socket_epoll_add(epoll, pidfd, 4);
```

Comment:
These errs are not checked?

Reply:
Also for single fd, poll is more efficient, but not big deal

Reply:
The branch with private ECR support does this without epoll, but this was added in a commit that also handles piping to a child's `stdin`.